CDCgov/prime-reportstream
# 1345 Add end to end test with data comparison

This PR adds a new CLI test called "compare" that reads a configuration file for test data, sends the test data to the RS API, then reads the lineage results and compares the resulting uploaded data from RS to the fake SFTP site and returns test results.  This test runs similar to the end2end test with the exception of comparing the resulting data at the end.

**LIMITATIONS:** Only supports CSV and HL7_BATCH receivers.  Tests are run sequentially.  Will open another issue to add support for HL7 single messages like for FL (did not add it in this PR to keep review smaller).

**Configuration file:**
Specifies what the input and expected reports are for the test.
  compare-test-config.csv has the following columns:
    Input File,Sender,Expected File,Organization name,Receiver name,Expected count
  Where:
    Input File = input HL7 or CSV file relative to the config file folder
    Sender = the name of the sender including organization (e.g. ignore.ignore-hl7)
    Expected File = the name of one of the expected output files relative to the config file folder
    Organization name = the name of the receiver organization (e.g. ignore)
    Receiver name = the name of the receiver (e.g. CSV)
    Expected count = The number of expected records sent as an integer
  You can specify one or more expected files for a given input file by having multiple lines with the same input
  file.  E.g.
    Input File,Sender,Expected File,Organization name,Receiver name,Expected count
    test-0001-input-covid-19.hl7,ignore.ignore-hl7,test-0001-az-covid-19-hl7.hl7,ignore,HL7_BATCH,5
    test-0001-input-covid-19.hl7,ignore.ignore-hl7,test-0001-pima-az-covid-19.csv,ignore,CSV,5

To test:
Run the following command: `./prime test --run=compare` and verify the tests run successfully.  

## Changes
- Added compare test to CLI
- Added some unit tests to make sure the comparison works as expected instead of relying on manual testing.

## Checklist

### Testing

### Process

## Fixes

## To Be Done
- Update the HL7 and CSV conversion tests to use a config file to support multiple schemas, plus use the CompareData functions to compare instead of the ones in their files.
- Add support for HL7 single message files to this test
